include/crocoddyl/multibody/actions/constraint-fwddyn.hxx

```diff
@@ -0,0 +1,196 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2021, INRIA
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#if PINOCCHIO_VERSION_AT_LEAST(2, 9, 0)
+
+namespace crocoddyl {
+
+template <typename Scalar>
+DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::DifferentialActionModelConstraintFwdDynamicsTpl(
+    boost::shared_ptr<StateMultibody> state, boost::shared_ptr<ActuationModelAbstract> actuation,
+    const PINOCCHIO_STD_VECTOR_WITH_EIGEN_ALLOCATOR(RigidConstraintModel) & contacts,
+    boost::shared_ptr<CostModelSum> costs, const ProximalSettings& settings)
+    : Base(state, actuation->get_nu(), costs->get_nr()),
+      actuation_(actuation),
+      contacts_(contacts),
+      costs_(costs),
+      pinocchio_(*state->get_pinocchio().get()),
+      settings_(settings) {
+  if (costs_->get_nu() != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "Costs doesn't have the same control dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  Base::set_u_lb(Scalar(-1.) * pinocchio_.effortLimit.tail(nu_));
+  Base::set_u_ub(Scalar(+1.) * pinocchio_.effortLimit.tail(nu_));
+}
+
+template <typename Scalar>
+DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::~DifferentialActionModelConstraintFwdDynamicsTpl() {}
+
+template <typename Scalar>
+void DifferentialActionModelConstraintFwdDynamicsTpl<Scalar>::calc(
+    const boost::shared_ptr<DifferentialActionDataAbstract>& data, const Eigen::Ref<const VectorXs>& x,
+    const Eigen::Ref<const VectorXs>& u) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+  if (static_cast<std::size_t>(u.size()) != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "u has wrong dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  DifferentialActionDataConstraintFwdDynamicsTpl<Scalar>* d =
+      static_cast<DifferentialActionDataConstraintFwdDynamicsTpl<Scalar>*>(data.get());
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> q = x.head(state_->get_nq());
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> v = x.tail(state_->get_nv());
+
+  // Computing the forward dynamics with the holonomic constraints defined by the contact model
+
+  actuation_->calc(d->multibody.actuation, x, u);
+  // TODO: Add mu
+  d->xout = pinocchio::constraintDynamics(pinocchio_, d->pinocchio, q, v, d->multibody.actuation->tau, contacts_,
```

Comment:
Should we call `pinocchio::computeCentroidalMomentum` for cost function that depends on centroidal quantities?